TimeTracking/TimeTracking.php

```diff
@@ -20,239 +20,266 @@
+    }
+
+    function hooks() {
+        return array(
+            'EVENT_LAYOUT_RESOURCES' => 'resources',
+            'EVENT_VIEW_BUG_EXTRA' => 'ev_view_bug',
+            'EVENT_MENU_ISSUE'     => 'timerecord_menu',
+            'EVENT_MENU_MAIN'      => 'showreport_menu',
+            'EVENT_VIEW_BUGNOTE'   => 'ev_view_bugnote',
+            'EVENT_BUGNOTE_ADD_FORM' => 'ev_bugnote_add_form',
+            'EVENT_BUGNOTE_DATA'   => 'ev_bugnote_add_validate',
+            'EVENT_BUGNOTE_ADD'    => 'ev_bugnote_added',
+            'EVENT_VIEW_BUG_DETAILS' => 'ev_view_bug_details',
+            'EVENT_LAYOUT_CONTENT_BEGIN' => 'ev_layout_content_begin',
+            'EVENT_FILTER_COLUMNS' => 'ev_register_columns',
+        );
+    }
+
+    /**
+     *
+     */
+    function config() {
+        return self::getConfig();
+    }
+
+    /**
+     *
+     */
+    function init() {
+        plugin_require_api( 'core/timetracking_api.php' );
+        plugin_require_api( 'core/stopwatch_api.php' );
+        plugin_require_api( 'core/columns.php' );
+        plugin_require_api( 'core/reports.php' );
+    }
+
+    /**
+     *
+     */
+    function errors() {
+        return array(
+            TimeTracking\ERROR_INVALID_TIME_FORMAT => plugin_lang_get( 'ERROR_INVALID_TIME_FORMAT' ),
+            TimeTracking\ERROR_ID_NOT_EXISTS => plugin_lang_get( 'ERROR_ID_NOT_EXISTS' ),
+        );
+    }
+
+    /**
+     *
+     */
+    function resources() {
+        $res  = '<link rel="stylesheet" type="text/css" href="'. plugin_file( 'timetracking.css' ) .'"/>';
+        $res .= '<script type="text/javascript" src="'. plugin_file( 'timetracking.js' ) .'"></script>';
+        $res .= '<script type="text/javascript" src="'. plugin_page( 'javascript_translations' ) .'"></script>';
+        if( \TimeTracking\stopwatch_enabled() ) {
+            $res .= '<script type="text/javascript" src="'. plugin_file( 'stopwatch.js' ) .'"></script>';
+        }
+        return $res;
+    }
+
+    /**
+     *
+     */ 
+    function ev_register_columns( $p_event ) {
+        return array(
+            new TimeTracking\ColumnTotalTime(),
+            new TimeTracking\ColumnMyTime(),
+        );
+    }
+
+    /**
+     * Show time tracking info within the bugnote activity area
+     */
+    function ev_view_bugnote( $p_event, $p_bug_id, $p_note_id, $p_is_private ) {
+        $t_record = TimeTracking\get_record_for_bugnote( $p_note_id );
+        if( !$t_record ) {
+            return;
+        }
+        if( TimeTracking\user_can_view_record_id( $t_record['id'] ) ) {
+            TimeTracking\print_bugnote_label_row( $t_record, $p_is_private );
+        }
+    }
+
+    /**
+     * Prints the time tracking inputs within the bugnote-add form
+     * @param type $p_event
+     * @param type $p_bug_id
+     */
+    function ev_bugnote_add_form( $p_event, $p_bug_id ) {
+        $t_enabled_on_bugnote_add_form = plugin_config_get('enabled_on_bugnote_add_form');
```

Comment:
Note that config options can be overridden for specific projects and/or users.
I see value at setting this option individually for some projects. Even if the UI does not support, this should be supported here.

Should we add explicit project from the bug_id as `plugin_config_get` parameter?
Omitting them, will relay to config_get for current user & project overrides, but:
Is there a scenario where this form is rendered for a bug_id, with a different "current project" than the bug's project?

Reply:
OK, going to do refactor